tests_and_analysis/performance_benchmarking/artifact_handling/copy_benchmark_artifacts.py

```diff
@@ -0,0 +1,228 @@
+        "-r", "--range", action="store", nargs=2, dest="range", type=int,
+        help="The range of Jenkins builds you wish to recover artifacts from."
+    )
+    return parser
+
+
+def coerce(value: int, lower_bound: int, upper_bound: int) -> int:
+    """
+    Coerce the value into between the lower bound and upper bound (inclusive).
+
+    Parameters
+    ----------
+    value : int
+        The value to coerce
+    lower_bound : int
+        The inclusive lower bound to coerce the value into
+    upper_bound : int
+        The inclusive upper bound to coerce the value into
+
+    Returns
+    -------
+    int
+        The value coerced into being between the lower and upper bound.
+    """
+    return max(lower_bound, min(value, upper_bound))
+
+
+def coerce_range(performance_benchmarking_response_json: Dict,
+                 range_parsed: Union[None, List[int]]) -> Tuple[int, int]:
+    """
+    Take details of the builds and find the first and last build numbers.
+    Coerce the range parsed by the user into between those first and last build
+    numbers.
+
+    Parameters
+    ----------
+    performance_benchmarking_response_json : Dict
+        A json response from Jenkins with build details including firstBuild
+        and lastBuild
+    range_parsed : Union[None, List[int]]
+        If no range was parsed this will be None, if a range has been parsed
+        this will be a list of two integers.
+
+    Returns
+    -------
+    Tuple[int, int]
+        The first value is the range start point and the second is the range
+        end point.
+    """
+    # The first and last build numbers
+    largest_possible_range = (
+        performance_benchmarking_response_json["firstBuild"]["number"] - 1,
+        performance_benchmarking_response_json["lastBuild"]["number"]
+    )
+    if range_parsed:
+        # Force both build numbers to be within the boundaries
+        lower_build_number = coerce(
+            range_parsed[0],
+            largest_possible_range[0], largest_possible_range[1]
+        )
+        upper_build_number = coerce(
+            range_parsed[1],
+            largest_possible_range[0], largest_possible_range[1]
+        )
+        return lower_build_number, upper_build_number
+    else:
+        # When there is no range parsed copy all the builds
+        return largest_possible_range
+
+
+def write_to_file(directory: str, filename: str,
+                  performance_benchmarks_json: str):
+    # Check access
+    test_file = os.path.join(directory, "test.txt")
+    try:
+        open(test_file, "w+")
+        # We have access therefore write
+        # No need to copy if the file already exists
+        if not os.path.exists(directory):
+            os.mkdir(directory)
+            filepath = os.path.join(directory, filename)
+            json.dump(performance_benchmarks_json, open(filepath, "w+"))
+    except FileNotFoundError:
+        print("Cannot access {}. Exiting.".format(directory))
```

Comment:
Also, because checking if the directory exists and creating it doesn't happen until after the test file is attempted to be created, this fails if the directory doesn't exist, even if we have the permissions. But not creating the test file should fix this anyway.

Reply:
Yes, this was what I thought would be logical as well. However, when copying to network locations (at least on Windows) if we do not have permissions to access it the script hangs on os.path.exists and won't complete or fail. Hence the creating of a test file.

Reply:
Not that I'm sure it's the best way around it, but I tried a few different methods and this was the one I found that worked most reliably.

Reply:
Do we need to check if the directory exists before creating it? If `os.mkdir` is called and the directory already exists a `FileExistsError` is raised, which we can catch. Or does `os.mkdir` have the potential to hang also?